### PR TITLE
Support partitioning db by block height 1395

### PR DESCRIPTION
**Pull request by lzxm160**
work on #1395

Block header,body ,footer and receipt map to different dbs according to block height,
other datas still store in default db,chain.db.
The default split length is 0，it means don't need to split db.
The split height config means it will split to diff db since that height.

If split length is 100000,every db's size is about 400M.
The new db's name like this:
chain-1.db
chain-2.db

### CODE REVIEW

File: blockchain/blockdao.go [hunk 1]
```diff
@@ -700,3 +822,11 @@ func deleteActions(dao *blockDAO, blk *block.Block, batch db.KVStoreBatch) error
 
 	return nil
 }
+
+// getDBIndex get db index from block height
+func getDBIndex(hei uint64, split uint64) int {
+	if split == 0 {
```
Comment (zjshen14) [inline blockchain/blockdao.go]: Is it to support when SplitDBLength is not set (default 0), all data will be written into the same db?

File: blockchain/blockdao.go [hunk 2]
```diff
@@ -434,9 +447,20 @@ func (dao *blockDAO) putBlock(blk *block.Block) error {
...
+	whichDB := getDBIndex(blk.Height(), dao.cfg.SplitDBLength)
+	kv, err := dao.getDBFromIndex(whichDB)
+	if err != nil {
+		return err
+	}
+	err = kv.Commit(batchForBlock)
+	if err != nil {
+		return err
+	}
+	blockHashDBKey := append(blockHashDBPrefix, hash[:]...)
```
Comment (zjshen14) [inline blockchain/blockdao.go]: When lookup by blockHash, can we just lookup for height first? We will know which db to reach as long as the height is known.

File: blockchain/blockdao.go [hunk 3]
```diff
@@ -606,6 +652,82 @@ func (dao *blockDAO) deleteTipBlock() error {
...
+	cfg := dao.cfg
+	withSuffix = path.Base(cfg.DbPath)
+	suffix = path.Ext(withSuffix)
+	name = strings.TrimSuffix(withSuffix, suffix)
+	name += fmt.Sprintf("-%d", index) + ".db"
+	cfg.DbPath = path.Dir(cfg.DbPath) + "/" + name
+
+	kvstore = db.NewBoltDB(cfg)
+	dao.kvstores.Store(index, kvstore)
+	err = kvstore.Start(context.Background())
```
Comment (zjshen14) [inline blockchain/blockdao.go]: A follow up may be closing the keystore when it's not been used for a while

File: blockchain/blockdao.go (hunk 1 above)
Comment (lzxm160) [reply blockchain/blockdao.go]: sure,now default is 100000,I'll change to 0

File: blockchain/blockdao.go (hunk 2 above)
Comment (lzxm160) [reply blockchain/blockdao.go]: see,I'll try delete this key

File: blockchain/blockdao.go (hunk 3 above)
Comment (lzxm160) [reply blockchain/blockdao.go]: I  think it will close when call dao.lifecycle.Stop()

**Comment by lzxm160** [comment]
> And think of the compatibility or migration plan. Currently all the blocks are stored in the one db, which couldn't be found with you new height->db mapping

I'll think about it

**Pull request closed without merging**